src/C2_economic_functions.py

```diff
@@ -76,74 +76,148 @@ def crf(project_life, discount_factor):
-        # Very first investment is in year 0
-        if count_of_replacements == 0:
-            capex = first_time_investment
-        else:
-            # replacements taking place in year = number_of_replacement * lifetime
-            if count_of_replacements * lifetime != project_life:
-                capex = capex + first_time_investment / (
-                    (1 + discount_factor) ** (count_of_replacements * lifetime)
-                )
-
-    # Subtraction of component value at end of life with last replacement (= number_of_investments - 1)
-    if number_of_investments * lifetime > project_life:
-        last_investment = first_time_investment / (
-            (1 + discount_factor) ** ((number_of_investments - 1) * lifetime)
+        number_of_investments = int(
+            round((project_lifetime + age_of_asset) / asset_lifetime + 0.5)
         )
+
+    replacement_costs = 0
+
+    # In case the asset_lifetime is larger then
+    latest_investment = first_time_investment
+    # Starting from first investment (in the past for installed capacities)
+    year = -age_of_asset
+
+    present_value_of_capital_expenditures = pd.Series([latest_investment], index=[year])
+
+    # Looping over replacements, excluding first_time_investment in year (0 - age_of_asset)
+    for count_of_replacements in range(1, number_of_investments):
+        # replacements taking place after an asset ends its lifetime
+        year += asset_lifetime
+
+        # Update latest_investment (to be used for residual value)
+        latest_investment = first_time_investment / ((1 + discount_factor) ** (year))
+        # Add latest investment to replacement costs
+        replacement_costs += latest_investment
+        # Update cash flow projection (specific)
+        present_value_of_capital_expenditures[year] = latest_investment
+
+    # Calculation of residual value / value at project end
+    year += asset_lifetime
+    if year > project_lifetime:
         # the residual of the capex at the end of the simulation time takes into
+        linear_depreciation_last_investment = latest_investment / asset_lifetime
         # account the value of the money by deviding by (1 + discount_factor) ** (project_life)
-        linear_depreciation_last_investment = last_investment / lifetime
-        capex = capex - linear_depreciation_last_investment * (
-            number_of_investments * lifetime - project_life
-        ) / (1 + discount_factor) ** (project_life)
+        value_at_project_end = (
+            linear_depreciation_last_investment
+            * (year - project_lifetime)
+            / (1 + discount_factor) ** (project_lifetime)
+        )
+        # Subtraction of component value at end of life with last replacement (= number_of_investments - 1)
+        replacement_costs -= value_at_project_end
+        # Update cash flow projection (specific)
+        present_value_of_capital_expenditures[project_lifetime] = -value_at_project_end
```

Comment:
This is not used, no?

Reply:
True, not yet. But it is something that the end users would like to see, so I wanted to lay the groundwork while I knew how to.